Overwrite

yohplala opened the pull request:
Hi @martindurant 

I am opening this new PR in replacement of PR #546.
Following your proposal, I setup most (all actually?) modifications of a DataFrame in case of data overlap outside fastparquet `write` function. 
More precisely, definition of the metric to sort and identify overlap (by date), retrieval of overlapped data, concatenation with new one, and removal of duplicates and sorting steps are all removed.

I have only kept an 'overwrite' mode.

The use case is as follow:
- data is already stored, with `partition_on` listing one or several columns AND `row_group_offsets=0`.
- when overwrite is used, for the same combination of values from columns listed in `partition_on`, if a file is already existing, it is overwritten
- its metadata is also replaced.

I did not create any new parameter for this `overwrite`. Instead I used a specific combination of existing parameters... but likely hard to be found by anyone :) so worth reading the docstring to be aware of it :)
This combination makes use of a combination that was actually raising a bug (when `row_group_offsets = 0`).

This function is triggered with use of `write` function and a specific combination of parameters.
row_group_offsets = 0
file_scheme = 'hive'
partition_on initialized
append = True

A test case is provided in the PR.
Thanks in advance for your feedback about it.
Have a good evening,
Bests,

PS:
 - I spot a bug that I did not solve. I am opening a ticket. ('drill' cannot be used with 'append').
 - I am wondering as it is my case. Do you think we can save any CPU by providing the `fp.ParquetFile`as a parameter to `write`in case of `append` to initialize `pf` in place of the line [942](https://github.com/dask/fastparquet/blob/d9cb86d33cedb7cdda1893e569c48ced75e419d0/fastparquet/writer.py#L942)
 In my case, because I will have managed concatenation, I will have it, so I can provide it through a parameter when calling `fastparquet.write()`. Do you think it does worth it?

martindurant commented:
> I spot a bug that I did not solve. I am opening a ticket. ('drill' cannot be used with 'append').

Yes, I spotted this while looking through your work too. It happens because the assumed categorical column names in drill are "dir0", "dir1"..., but the pandas metadata has the original field names. Those original names should be applied, if you have them.

yohplala commented:
> > I spot a bug that I did not solve. I am opening a ticket. ('drill' cannot be used with 'append').
> 
> Yes, I spotted this while looking through your work too. It happens because the assumed categorical column names in drill are "dir0", "dir1"..., but the pandas metadata has the original field names. Those original names should be applied, if you have them.

Ah, I have not spot the same reason.
The trouble I have spotted lies line [943](https://github.com/dask/fastparquet/blob/d9cb86d33cedb7cdda1893e569c48ced75e419d0/fastparquet/writer.py#L943)
Value 'drill' is missing.
`if pf.file_scheme not in ['hive', 'empty', 'flat']:`

martindurant commented:
Ah, so maybe we were just avoiding the case...

yohplala commented:
> Ah, so maybe we were just avoiding the case...

Hmm, just 2 lines above, 'drill' is checked. The 'full' extract actually reads:
```python
    elif file_scheme in ['hive', 'drill']:
        if append:
            pf = api.ParquetFile(filename, open_with=open_with)
            if pf.file_scheme not in ['hive', 'empty', 'flat']:
                raise ValueError('Requested file scheme is %s, but '
                                 'existing file scheme is not.' % file_scheme)
```

So the 1st line lets you enter if you have 'drill', and then raises an error.

martindurant commented:
As usual, the best thing would be to have a test for this...
I'm not sure people use this Drill scheme anymore. It wouldn't surprise me if Drill itself supported the hive style by now. Clearly, it stopped being important for the purposes of fastparquet...

yohplala commented:
> As usual, the best thing would be to have a test for this...
> I'm not sure people use this Drill scheme anymore. It wouldn't surprise me if Drill itself supported the hive style by now. Clearly, it stopped being important for the purposes of fastparquet...

Hi @martindurant 
Well I am surprised of this feedback. I provided a test in file:
https://github.com/yohplala/fastparquet/blob/overwrite/fastparquet/test/test_overwrite.py

Yes, I remeber I cleaned a line regarding drill. I can remove this if required. I tried to use it, but saw that it is not keeping column name, so I drop to use it.

Please, could you clarify what is wrong with the test case provided?

martindurant commented:
Looking back at this now.
Would you say this is a correct summary of this new codepath:
In the case that there is partitioning (but only one file per partition), *replace* any partition for which we provide data, but leave those for which we don't.

This is a precursor for *append* data to partition by growing row-groups (optional drop duplicates).

I think I've been caught up on the wording - it's more in the text, than in the code.

yohplala commented:
> Looking back at this now.
> Would you say this is a correct summary of this new codepath:
> In the case that there is partitioning (but only one file per partition), _replace_ any partition for which we provide data, but leave those for which we don't.
> 
> This is a precursor for _append_ data to partition by growing row-groups (optional drop duplicates).
> 
> I think I've been caught up on the wording - it's more in the text, than in the code.

Hi @martindurant.

I agree with your wording. If I may, to be even more specific (in case user does not know what is behing the word `partition`).
_replace_ the content of any _partition folder_ for which we provide data, but leave those for which we don't.

The start of your sentence raises a _wrong_ understanding:
> In the case that there is partitioning (but only one file per partition)
The user may understand that an exception will be raised. But no verification on the number of files is made.
So what does happen in this case if:
- 1st I write data in a parition with several files (`row_group_offsets != 0`)
- 2nd, then I write new data in same partition, this time with `row_group_offsets == 0`

The feature only overwrite `part.0.parquet`, so data in the other files will remain.
`fmd`should still be consistent, as this feature does not remove `rg` from `rgps`.

I am not advertizing on this mode becase AFAIK, user has no real way to know what data is in `part.0.parquet` that will get overwritten. So no real point in using such a case according to me.

martindurant commented:
I think that overwriting one file in a partition directory where there are potentially more files is not acceptable - users would not expect it to behave like this. In such a situation, it would make sense to, at a minimum, make sure of only one file per directory else raise exception; or change the logic to remove all the files in said directory first (also in the metadata).

yohplala commented:
> I think that overwriting one file in a partition directory where there are potentially more files is not acceptable - users would not expect it to behave like this. In such a situation, it would make sense to, at a minimum, make sure of only one file per directory else raise exception; or change the logic to remove all the files in said directory first (also in the metadata).

Hi @martindurant ,
Agreed, I will look into your proposals this week-end. I'll keep you informed.

yohplala commented:
Hi @martindurant 
I have implemented the check for raising an error in case several part files are a partition folder.

Regarding the checks, I am confused, it does not seem to be related to my PR.
The very 1st fail regarding absence of warning when importing fastparquet.

The 2 warnings apparently issued do not appear related to this PR:
```
=========================== short test summary info ============================
FAILED fastparquet/test/test_api.py::test_import_without_warning - subprocess...
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
======================== 1 failed, 3 warnings in 2.83s =========================
```
```
DeprecationWarning: `np.long` is a deprecated alias for `np.compat.long`. To silence this warning, use `np.compat.long` by itself. In the likely event your code does not need to work on Python 2 you can use the builtin `int` for which `np.compat.long` is itself an alias. Doing this will not modify any behaviour and is safe. When replacing `np.long`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
```
```
FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
    import pandas.util.testing
```
Am I understanding correctly?

martindurant commented:
This will be due to the new version of numpy, jus released. It would be worthwhile correcting our code to avoid the warnings, but you could pin the version of numpy for the purposes of testing for now.
For the pandas one, I would explicitly ignore the warning where that import happens.

yohplala commented:
> This will be due to the new version of numpy, jus released. It would be worthwhile correcting our code to avoid the warnings, but you could pin the version of numpy for the purposes of testing for now.
> For the pandas one, I would explicitly ignore the warning where that import happens.

As far as I understand, this has to be done in the project settings, and I have absolutely no knowledge about that.
Do you know which file I should modify?
(it is not in the code I have been working on)

martindurant commented:
Actually, try merging from master, we may have removed this check about warnings.

On February 14, 2021 1:49:06 PM EST, yohplala <notifications@github.com> wrote:
>> This will be due to the new version of numpy, jus released. It would
>be worthwhile correcting our code to avoid the warnings, but you could
>pin the version of numpy for the purposes of testing for now.
>> For the pandas one, I would explicitly ignore the warning where that
>import happens.
>
>As far as I understand, this has to be done in the project settings,
>and I have absolutely no knowledge about that.
>Do you know which file I should modify?
>(it is not in the code I have been working on)
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/dask/fastparquet/pull/553#issuecomment-778822854

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

yohplala commented:
> Actually, try merging from master, we may have removed this check about warnings.
> […](#)

I did it before pushing the update. I checked, there is no new commit to main since then.

martindurant commented:
Now you can merge from master :)

martindurant commented:
I mean, merge from "main" - the default branch has been renamed!

yohplala commented:
> I mean, merge from "main" - the default branch has been renamed!

Thanks @martindurant . All seems ok now :)

yohplala commented:
> In such a situation, it would make sense to, at a minimum, make sure of only one file per directory else raise exception

@martindurant 
You can check the exception aised with following test.
I did not include it in the test file because I don't know how to check that specifically this exception is raides. Should I investigate?

```python
import os
import pandas as pd
import fastparquet

path = os.path.expanduser('~/Documents/code/data/')
file = path + 'test_parquet'

df1 = pd.DataFrame({'humidity': [0.3, 0.8, 0.9, 0.7],
                    'pressure': [1e5, 1.1e5, 0.95e5, 1e5],
                    'location': ['Paris', 'Paris', 'Milan', 'Paris'],
                    'exterior': ['yes', 'no', 'yes', 'yes']})

fastparquet.write(file, df1, row_group_offsets = 1, file_scheme='hive',
                  write_index=False,partition_on=['location', 'exterior'])

fastparquet.write(file, df1, row_group_offsets = 0, file_scheme='hive',
                  write_index=False,partition_on=['location', 'exterior'], append=True)

```

yohplala commented:
Hello @martindurant 
Please, is this PR ok? Interested to have it merged if it is ok of course.
Thanks for your feedbacks, have a good day, bests,

martindurant commented:
A test in test_overwrite.py might look like

```python
def test_overwrite_ex(tempdir):
    df1 = pd.DataFrame({'humidity': [0.3, 0.8, 0.9, 0.7],
                    'pressure': [1e5, 1.1e5, 0.95e5, 1e5],
                    'location': ['Paris', 'Paris', 'Milan', 'Paris'],
                    'exterior': ['yes', 'no', 'yes', 'yes']})

    fastparquet.write(file, df1, row_group_offsets = 1, file_scheme='hive',
                  write_index=False,partition_on=['location', 'exterior'])

    with pytest.raises(ValueError, match="^Some partition folders"):
        fastparquet.write(file, df1, row_group_offsets = 0, file_scheme='hive',
                  write_index=False,partition_on=['location', 'exterior'], append=True)
```

martindurant commented on fastparquet/writer.py:
```diff
@@ -814,15 +815,27 @@ def write(filename, data, row_group_offsets=50000000,
-    """ Write Pandas DataFrame to filename as Parquet Format
+    """ Write Pandas DataFrame to filename as Parquet Format.
+
+        3 writing modes can be identified:
```
Please relegate this block to the `append` parameter; potentially take it out completely and put it in the docs. The reason I say this, is that the matter is hard to understand, and by trying to fit it here, it becomes rather terse. In the docs, you could add examples to show what overwrite and append modes are supposed to do.

martindurant commented on fastparquet/writer.py:
```diff
@@ -962,17 +996,41 @@ def write(filename, data, row_group_offsets=50000000,
                 )
-                fmd.row_groups.extend(rgs)
+                if not exist_rgps:
+                    # Append or 'standard' write mode.
```
please call them "append" and "overwrite in place"

martindurant commented on fastparquet/writer.py:
```diff
@@ -814,15 +815,27 @@ def write(filename, data, row_group_offsets=50000000,
+
+        3 writing modes can be identified:
+            - create file(s),
+            - append data by creating new files without removing existing ones,
```
So the first is `append=False`; the other two are `append=True` but depend on the other arguments? I would call these modes "append" and "overwrite-in-place". 

Really, it may make sense to have `append=True|False|overwrite`, and ensure that if the third is chose, `row_group_offsets==0`.

yohplala replied:
You are right. This came to my mind as well, but as to enable it, you need a specific combination of parameters, this new value of `append` only confirms this 'overwrite-in-place' mode. Ok, doing it.

martindurant reviewed (commented):
The code is fine, but I still think that the API is confusing. I have commented to try to exmplain my unease.

yohplala commented:
Hi @martindurant, thanks for your continued support.
[x] test of exception is ok
[x] overwrite mode has now to be enabled by use of `append='overwrite'` in addition of using a specific combination of parameters ('hive', row_group_offsets = 0, and use of `partition_on`)
[x] I added a line to mention the overwrite mode in 'quickstart'
[x] I moved the block listing the 3 uses of write in the 'append' part

Please, let me know if further adjustements are needed.
I believe that this way, thanks to your feedbacks, the content of this PR has been adjusted to the right perimeter, thank you.
Bests,

martindurant commented on fastparquet/writer.py:
```diff
@@ -902,10 +890,15 @@ def write(filename, data, row_group_offsets=50000000,
+        - If False, construct data-set from scratch; if True, add new row-group(s)
+          to existing data-set. In the latter case, the data-set must exist,
+          and the schema must match the input data.
+        - If 'overwrite', existing data can be replaced. To enable this, these other
```
"If 'overwrite', existing partitions will be replaced in-place, where the given data has any rows within a given partition"

yohplala replied:
done

martindurant commented on fastparquet/writer.py:
```diff
@@ -902,10 +890,15 @@ def write(filename, data, row_group_offsets=50000000,
+          to existing data-set. In the latter case, the data-set must exist,
+          and the schema must match the input data.
+        - If 'overwrite', existing data can be replaced. To enable this, these other
+          parameters have to be set to specific values:
```
" have to be set to specific values, or will raise ValueError:"

yohplala replied:
done

martindurant commented on docs/source/quickstart.rst:
```diff
@@ -67,6 +67,7 @@ Further options that may be of interest are:
 - the compression algorithms (typically "snappy", for fast, but not too space-efficient), which can vary by column
 - the row-group splits to apply, which may lead to efficiencies on loading, if some row-groups can be skipped. Statistics (min/max) are calculated for each column in each row-group on the fly.
 - multi-file saving can be enabled with the ``file_scheme`` keyword: hive-style output is a directory with a single metadata file and several data-files.
+- complementing ``file_scheme='hive'``, use of ``partition_on`` spread data in different directories. In this situation, ``append`` can be set to one value among ``False|True|'overwrite'``. With this latter mode, new data written in an existing directory replaces the existing content. 
```
Edit previous bullet:
- multi-file saving can be enabled with the ``file_scheme="hive"|"drill"``: directory-tree-partitioned output with a single metadata file and several data-files, one or more per leaf directory. The values used for partitioning are encoded into the paths of the data files.

and change this one
- append to existing data sets with ``append=True``, adding new row-groups. For the specific case of "hive"-partitioned data and one file per partition, ``append="overwrite"`` is also allowed, which replaces partitions of the data where new data are given, but leaves other existing partitions untouched.

yohplala replied:
done

martindurant reviewed (commented):
Almost! I have wording suggestions to try to make this clearer still.

yohplala commented:
Dear Martin, I hope this will do the trick.
Thanks for taking so much time, I appreciate your feedbacks.
Bests

martindurant commented:
Done! Thanks for your patience...
